src/composition/runner.rs

```diff
@@ -243,40 +267,101 @@ impl SubtaskHandleStream for SubgraphWatchers {
                     }) {
-                        let sender = sender.clone();
-                        let subgraph_name_c = subgraph_name.clone();
-                        let messages_abort_handle = tokio::spawn(async move {
-                            while let Some(change) = messages.next().await {
-                                let _ = sender
-                                    .send(SubgraphChanged {
-                                        name: subgraph_name_c.to_string(),
-                                        sdl: change.sdl().to_string(),
-                                    })
-                                    .tap_err(|err| tracing::error!("{:?}", err));
-                            }
-                        })
-                        .abort_handle();
-                        let subtask_abort_handle = subtask.run();
-                        abort_handles.insert(
-                            subgraph_name.to_string(),
-                            (messages_abort_handle, subtask_abort_handle),
-                        );
+                        // If a SchemaSource::Subgraph or SchemaSource::Sdl was added, we don't
+                        // want to spin up watchers; rather, we emit a SubgraphSchemaChanged event with
+                        // either what we fetch from Studio (for Subgraphs) or what the SupergraphConfig
+                        // has for Sdls
+                        if let SubgraphWatcherKind::Once(non_repeating_fetch) =
+                            subgraph_watcher.watcher()
+                        {
+                            let _ = non_repeating_fetch
+                                .run()
+                                .await
+                                .tap_err(|err| {
+                                    tracing::error!("failed to get {subgraph_name}'s SDL: {err:?}")
+                                })
+                                .map(|sdl| {
+                                    let _ = sender
+                                        .send(SubgraphEvent::SubgraphChanged(
+                                            SubgraphSchemaChanged {
+                                                name: subgraph_name.to_string(),
+                                                sdl,
+                                            },
+                                        ))
+                                        .tap_err(|err| tracing::error!("{:?}", err));
+                                });
+                        // When we have a SchemaSource that's watchable, we start a new subtask
+                        // and add it to our list of subtasks
+                        } else {
+                            let (mut messages, subtask) =
+                                Subtask::<SubgraphWatcher, WatchedSdlChange>::new(subgraph_watcher);
+
+                            let sender = sender.clone();
+                            let subgraph_name_c = subgraph_name.clone();
```

Comment:
nit: `subtask_clone_c` feels like a bit of a smell. It might feel slightly cleaner if we reused the subgraph_name variable in a block (which seems to be an idiomatic rust thing). So like:
```
let messages_abort_handle = {
  let sender = sender.clone();
  let subgraph_name = subgraph_name.clone();
  tokio::spawn({
    ...
  }).abort_handle();
}
```